Mobile - Anticipate multiple forms of vaccine codes group names in immunizations

aherzberg opened the pull request:
## Summary
Instead of assuming the vaccine group is in the second vaccine code, it will now look for whichever vaccine code has the vaccine group in the display field and then fall back to the second and then first vaccine group. 

## Related issue(s)
https://github.com/department-of-veterans-affairs/va-mobile-app/issues/7527

## Testing done

- *Describe what the old behavior was prior to the change*
- *Describe the steps required to verify your changes are working as expected. Exclusively stating 'Specs run' is NOT acceptable as appropriate testing*
- *Describe the tests completed and the results*

## Screenshots
_Note: Optional_


## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas)*

## Acceptance criteria

- [ ]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

aherzberg commented on modules/mobile/app/models/mobile/v0/adapters/immunizations.rb:
This is being very defensive because I'm seeing multiple ways the vaccine group can come in the older VCRS and what is currently coming in on staging.

aherzberg commented on modules/mobile/spec/request/v1/immunizations_request_spec.rb:
```diff
@@ -32,43 +32,32 @@
       it 'matches the expected schema' do
         # TODO: this should use the matcher helper instead (was throwing an Oj::ParseError)
         # expect().to match_json_schema('immunizations')
```
Created a separate ticket to solve this issue. https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/7614

aherzberg replied:
I think it may be the other way around if I'm reading that ticket right. Even if we combine the spec schemas and the doc schemas, the specs will still need to call a file to validate, in which case, when that stub is happening with the RSA key, it won't work. So 7614 will need to be solved before all specs can validate a schema.

kpethtel replied:
Sorry, I didn't really read the ticket. I was just looking at the code and assumed you meant fix the schema. Didn't realize it was caused by an auth stubbing issue.

aherzberg commented on modules/mobile/spec/support/schemas/v1/immunizations.json:
can't be used until ticket linked above is solved but I made it already when trying to get it to work.

aherzberg commented on modules/mobile/spec/support/schemas/v1/immunizations.json:
can't be used till ticket linked above is done but I already made it when trying to figure out problem

aherzberg commented on spec/support/vcr_cassettes/mobile/lighthouse_health/get_immunizations.yml:
new call from staging. different format of vaccine codes

kpethtel reviewed (dismissed):
LGTM. It's a little hard to tell because the comment got removed, but I think you're saying you re-created the vcr which explains all the spec changes. A fine reminder that we do sometimes need to refresh our specs. 

I find it a little odd that many of the new records don't have dose series values, but it looks like at least one still does so I guess it works.

aherzberg commented on modules/mobile/app/models/mobile/v0/adapters/immunizations.rb:
```diff
@@ -51,9 +51,15 @@ def manufacturer(immunization, group_name)
+          filtered_codes = vaccine_codes[:coding].select { |v| v[:display].include?('VACCINE GROUP: ') }
+
+          if filtered_codes.empty?
+            group_name = vaccine_codes.dig(:coding, 1, :display) || vaccine_codes.dig(:coding, 0, :display)
```
the second one  to be more human readable as a title such as FLU while the first one would be something like "Influenza, seasonal, injectable, preservative free". We don't have a hard business rule that says there can only be two but the new changes does account for the possibility of there being more and will pick up whichever one has that VACCINE GROUP text and then fall back to the second and first if nothing is caught

aherzberg commented:
@kris

> LGTM. It's a little hard to tell because the comment got removed, but I think you're saying you re-created the vcr which explains all the spec changes. A fine reminder that we do sometimes need to refresh our specs.
> 
> I find it a little odd that many of the new records don't have dose series values, but it looks like at least one still does so I guess it works.

oh wow. good catch. they also changed the name of the field for the dose series. I've updated that as well

kpethtel commented on modules/mobile/app/models/mobile/v0/adapters/immunizations.rb:
```diff
@@ -96,7 +96,7 @@ def dose_series(protocol_applied)
           series = protocol_applied.first
 
-          series[:series_doses_positive_int] || series[:series_doses_string]
+          series[:series_doses_positive_int] || series[:series_doses_string] || series[:dose_number_string]
```
Every day we stray further...

kpethtel reviewed (approved):
Nice catch on the dose series. Looks like it was semi-broken and we didn't even know it yet. Caught it before QA did.

aherzberg commented on modules/mobile/spec/request/v1/immunizations_request_spec.rb:
```diff
@@ -168,46 +130,57 @@
+
+          covid_no_manufacturer_immunization = response.parsed_body['data'].select do |i|
+            i['id'] == 'I2-LJAZCGMN3BZVQVKQCVL7KMTHJA000000'
+          end
```
This is what was causing the issue. for some reason, on github trying to get the exact page with size of one was causing an authentication error. I'm pretty confident this is a CI bug, not a bug in the code.

Tonksthebear reviewed (approved):
LGTM after the explanation in Kris's comments
